### ISSUE TITLE
Wanted: plug and play + dependency diffing

### ISSUE THREAD

**Issue by jeroenptrs** [denoserverless/darn#5]
## Functionality
Similar to Yarn PnP, darn should be able to locally keep a cache and validate/diff them when dependencies are reloaded/recompiled.

* Use a main cache for validated dependencies and a separate cache to fetch reloaded deps. Diff between the two (map dependencies to their individual hashes *and make sure this isn't slow!*) and notify the user on changes. If the user approves of changes and all checks pass (maybe performing test compiles/runs/tests), persist them to the 
* Keep this behind a flag in the cli and an option in the rc file, it should be opt-in for now
* We may need to introduce a dependency list and/or lockfile system to support this functionality

**Comment by bartlomieju** [comment]
Hi @jeroenptrs, I wanted to ask a few questions:


> * Use a main cache for validated dependencies and a separate cache to fetch reloaded deps. Diff between the two (map dependencies to their individual hashes _and make sure this isn't slow!_) and notify the user on changes. If the user approves of changes and all checks pass (maybe performing test compiles/runs/tests), persist them to the

Do you mean project dependencies, ie. 3rd party packages? If so, wouldn't this overlap with Deno itself which is responsible for downloading and caching of dependencies?

Or is it like darn is responsible for generating/updating `deps.ts` file, ie. allows to automatically add new package with specific version to `deps.ts` and reexport it from there. 

Anyway, looking forward to seeing how darn evolves! I'll try to contribute somehow

**Comment by jeroenptrs** [comment]
@bartlomieju

> Do you mean project dependencies, ie. 3rd party packages?

Yep! I was thinking about this use case for teams that want to share dependencies fast and easy, without having to re-fetch and also for an additional layer of security. E.g. if something that's 3rd party gets hijacked you get notified of changes for that dependency (what if you lock to a version but someone deletes the tag, adds malicious code and retags it, etc.).

> If so, wouldn't this overlap with Deno itself which is responsible for downloading and caching of dependencies?

 I'm still not sure about the best way to do this without wrapping something around Deno or overengineering stuff. One of the main reasons I'm considering it future functionality - but I'm definitely open to suggestions!

I was first thinking of creating a `.temp` folder and a `.cache` folder and pointing `$DENO_DIR/{deps|gen}` to them when running a command like `darn fetch` - which calls `deno fetch` in the background -, and doing the diffing between `.cache` (accepted dependencies) and `.temp` (reloaded dependencies).

One of the issues that springs to mind with this method is the actual deno executable loses it's reference since we use `$DENO_DIR/bin/deno`, unless you perform a symlink for that as well.

In short: the current ideas are still a bit messy 😜 

> Or is it like darn is responsible for generating/updating deps.ts file, ie. allows to automatically add new package with specific version to deps.ts and reexport it from there.

Yep! I was definitely thinking of something as a dependency list as the foundation for this functionality and then let whatever implementation gets decided in this issue build on top of that. And working with specific versions is definitely needed (this could be as high-level as a tag on github, or as low level as the hash of the dependency in the `$DENO_DIR/gen` folder.

**Comment by jeroenptrs** [comment]
Having a flag in `deno` to set the out dir at runtime (without having to do it with `DENO_DIR=/my/new/path` would be a welcome feature.

I'm building Darn with it being an *additional* tool in mind, so you wouldn't need to have it installed to run your code (just Deno). This feature - at the current moment - seems to force a change of perspective on that front.

**Comment by jeroenptrs** [comment]
> I'm still not sure about the best way to do this without wrapping something around Deno or overengineering stuff.

With this I meant wrapping *each version of deno in `$HOME/.darn/versions`.

But come to think of it, we might want to wrap `$DENO_DIR/bin/deno` for supporting something like automatically checking what deno version to use if there's a .darnrc present etc. And resolving to the actual deno executable in the background. In essence we could extend this functionality with mapping the `dep` and `gen` folders as well so you could do it entirely programmatically.

I'll update this issue's description once I've got something concrete

**Comment by bartlomieju** [comment]
Thanks for responses @jeroenptrs I can see value in proposed functionality, although I'm worried that `.darnrc.json` might be another `package.json` which is one of goals of Deno to get rid of it. Nevertheless I'm happy to learn more and contribute

**Comment by jeroenptrs** [comment]
> I'm worried that .darnrc.json might be another package.json which is one of goals of Deno to get rid of it

I think you can do both. In short (I'm on mobile), .darnrc.json should not be more than a runtime config (rc).

- on one side: if you run with darn, it'll point to the rc file
- on the other hand: running without darn should always work, aka _any project should be able to run without darn_.

So on one side if you want to use deno but i.e. want security in pinning dependencies and fast scripting without using npm/yarn, you could do that with darn 😁
